Replace legacy websocket provider with V2

fselmo opened the pull request:
### What was wrong?

- Replace the legacy WebSocket provider with V2.

We wanted to get these changes in before the first release of `v7` beta.

### How was it fixed?

- ``WebsocketProvider`` -> ``LegacyWebSocketProvider``
- ``WebsocketProviderV2`` -> ``WebSocketProvider``
- Update documentation to reflect the above changes

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

![20240206_142634](https://github.com/ethereum/web3.py/assets/3532824/efbf9921-b4c6-4323-b831-e426ecb89947)

kclowes commented on docs/quickstart.rst:
```diff
@@ -102,7 +104,7 @@ You can connect to a remote node by specifying the endpoint, just like the previ
    >>> w3 = AsyncWeb3(AsyncWeb3.AsyncHTTPProvider('https://<your-provider-url>'))
 
-   >>> w3 = Web3(Web3.WebsocketProvider('wss://<your-provider-url>'))
+   >>> w3 = await AsyncWeb3(AsyncWeb3.WebsocketProvider('wss://<your-provider-url>'))
```
There is a very good chance this is just me not knowing how to use the API, but I always get an error: `TypeError: object AsyncWeb3 can't be used in 'await' expression` when I run this. Heres the code for reference: 
```
>>> from web3 import AsyncWeb3
>>> w3 = await AsyncWeb3(
>>>     AsyncWeb3.WebsocketProvider('ws://localhost:8546')
>>> )
```

fselmo replied:
Nope, you're right. This sparked a whole discussion but to note here:

We can completely remove the need for `persistent_websocket()` by implementing the magic methods in `AsyncWeb3` itself. Those changes will come in here soon.

kclowes reviewed (approved):
Just putting this here so we don't forget... we need to add a deprecation warning (or some type of warning) in v6 on the `WebsocketProvider` to let people know about the change from `WebsocketProvider` -> `LegacyWebsocketProvider`, and provide the legacy alias. I'm also happy to add an issue, just let me know!

Looks good! I think there are a couple spots in the Quickstart docs where the new WebsocketProvider API doesn't work quite like the docs say it does, but that may just be user error :)

pacrob commented on web3/providers/websocket/websocket.py:
```diff
@@ -2,134 +2,243 @@
-from websockets.legacy.client import (
-    WebSocketClientProtocol,
+from websockets.exceptions import (
+    WebSocketException,
```
consistency nit - All the Providers are `Websocket...`, but the Exception (and a couple other places I found in search) use `WebSocket...` Probably good to make them all `Websocket`, yah?

pacrob commented:
The organization of providers feels unintuitive:

![image](https://github.com/ethereum/web3.py/assets/5199899/9915f4e2-1038-4f31-b1b2-13a76533dd58)

Maybe org the HTTP Providers together? Do we have future `PersistentConnectionProvider`s planned that would merit it staying as an umbrella term? If not, I feel like it's good to have the names of the actual providers users will be implementing show up first in docs. Can then include deets about `PersistentConnectionProvider` in more detailed description.

Just thinking out loud. What'chu think?

kclowes commented:
> The organization of providers feels unintuitive: ...

Yeah, we don't neatly split between Sync and Async providers any more. We really have Sync Providers, Async Providers, and Persistent Connection Providers, IMHO. Might be good to reflect that in the docs

reedsa commented on ethpm/backends/registry.py:
```diff
@@ -50,10 +50,10 @@ class RegistryURIBackend(BaseURIBackend):
         from web3 import (
             Web3,
-            WebsocketProvider,
+            WebsocketProviderLegacy,
```
What is required to make use of `WebsocketProvider` in all examples?

fselmo replied:
Good thought. `ethpm` is deprecated though and being removed in `v7` so I don't think it will be a good time investment here. If you see this somewhere else in the tests that would be good to note.

fselmo committed:
Establish ``WebsocketProviderV2`` as the principal WebSocket provider:

- ``WebsocketProvider`` -> ``LegacyWebsocketProvider``
- ``WebsocketProviderV2`` -> ``WebsocketProvider``
- Update documentation and code base to reflect these changes

Additionally:

- Remove the need for ``persistent_connection()`` on the ``AsyncWeb3`` class
  by implementing the magic methods on ``AsyncWeb3`` itself.
- Use a method wrapper for methods only allowed when ``AsyncWeb3`` is
  instantiated with a ``PersistentConnectionProvider``.

... fix lint

fselmo committed:
Factor out commonalities back to PersistentConnectionProvider

fselmo committed:
Clean up typing for persistent connection providers

fselmo committed:
Consolidate ``PersistentConnectionProvider`` docs; reorganize.

reedsa reviewed (approved):
lgtm!

fselmo committed:
stabilize is_connected() for async ipc

pacrob reviewed (approved):
More of a nit, but it's still bugging me that everything we get from the `websockets` lib uses capital S `WebSocket` and we use mostly `Websocket`. But as long as ya'll are good with it, lgtm :)

fselmo committed:
name change for more correctness: ``Websocket`` -> ``WebSocket``

fselmo commented:
> More of a nit, but it's still bugging me that everything we get from the `websockets` lib uses capital S `WebSocket` and we use mostly `Websocket`. But as long as ya'll are good with it, lgtm :)

Being that the protocol is WebSocket, let's make it happen since we're already breaking everything. Updated in the last commit.

fselmo committed:
newsfragment for #3225
